Adding server dependency information (#204)
temporalio/documentation: docs/server-dependencies.md

---
id: server-dependencies
title: Temporal server dependencies
sidebar_label: Dependencies
---

## Overview

This page details some of the requirements and dependencies needed to build and run an instance of Temporal.

All versions of Temporal require the following:

1. [Go 1.15+](https://golang.org/dl/)

> 2. [Go package dependencies](https://github.com/temporalio/temporal/blob/master/go.mod)

Reviewer: the url is pointing to the master branch
for each release, there is different set of dependencies
v1.2.2: https://github.com/temporalio/temporal/blob/v1.2.2/go.mod
v1.3.2: https://github.com/temporalio/temporal/blob/v1.3.2/go.mod

| 1.3.0            | 🟩 🟪 🟧 | 🟥     | 🟫              | 🟦        | 🟨          |
| 1.2.2            | 🟩 🟪 🟧 | 🟥     | 🟫              | 🟦        | 🟨          |
| 1.2.1            | 🟩 🟪 🟧 | 🟥     | 🟫              | 🟦        | 🟨          |
| 1.1.1            | 🟩 🟪    | 🟥     | 🟫              | 🟦        | 🟨          |
| 1.1.0            | 🟩 🟪    | 🟥     | 🟫              | 🟦        | 🟨          |
| 1.0.0            | 🟩 🟪    | 🟥     | 🟫              | 🟦        | 🟨          |

## How to use a different Temporal release version

To use a more recent version of Temporal, first [check to see](https://github.com/temporalio/temporal/releases) if an upgrade to the database schema is required. Newer binaries can not run with older database schemas. Some releases require changes to the schema, and some do not. If you are using a version that is older than 1.0.0, reach out to us at [community.temporal.io](http://community.temporal.io) to ask how to upgrade.

If you wish to use an older version of Temporal, you do not need to make schema changes as older binaries can operate with newer schemas. Note: we do not recommend using any version older than v1.0.0.

> We ensure that any consecutive versions are compatible in terms of database schema upgrades, however there is no guarantee that schema changes between *any* 2 non-consecutive versions are compatible. Please reach out to us or check the forums at [community.temporal.io](http://community.temporal.io) for more information.

Reviewer: `If you wish to use an older version of Temporal, you do not need to make schema changes as older binaries can operate with newer schemas.`
not really accurate, plz delete in favor of 
`We ensure that any consecutive versions are compatible in terms of database schema upgrades, however there is no guarantee that schema changes between *any* 2 non-consecutive versions are compatible.`

Reviewer: ```
We ensure that any consecutive versions are compatible in terms of database schema upgrades, however there is no guarantee that schema changes between *any* 2 non-consecutive versions are compatible.
```

the guarantee is more than just schema, but also new features and old behavior -> new behavior migrations

### Upgrade Cassandra schema

You can use the `temporal-cassandra-tool` to upgrade both the default and visibility schemas for your Cassandra DB:

**Example default schema upgrade:**

```
temporal_v1.2.1 $ temporal-cassandra-tool \
   --tls \
   --tls-ca-file <...> \
   --user <cassandra-user> \
   --password <cassandra-password> \
   --endpoint <cassandra.example.com> \
   --keyspace temporal \
   --timeout 120 \
   update \
   --schema-dir ./schema/cassandra/temporal/versioned

```

**Example visibility schema upgrade:**

```
temporal_v1.2.1 $ temporal-cassandra-tool \
   --tls \
   --tls-ca-file <...> \
   --user <cassandra-user> \
   --password <cassandra-password> \
   --endpoint <cassandra.example.com> \
   --keyspace temporal_visibility \
   --timeout 120 \
   update \
   --schema-dir ./schema/cassandra/visibility/versioned

```

### Upgrade MySQL / PostgreSQL schema

Use the `temporal-sql-tool`, which works similarly to the `temporal-cassandra-tool`.

> Refer to this [Makefile](https://github.com/temporalio/temporal/blob/v1.3.2/Makefile#L367-L383) for context.

Reviewer: sample commands:
```zsh
./temporal-sql-tool \
	--tls \
	--tls-enable-host-verification \
	--tls-cert-file <path to your client cert> \
	--tls-key-file <path to your client key> \
	--tls-ca-file <path to your CA> \ 
	--ep localhost -p 5432 -u temporal -pw temporal --pl postgres --db temporal update-schema -d ./schema/postgresql/v96/temporal/versioned
```

```zsh
./temporal-sql-tool \
	--tls \
	--tls-enable-host-verification \
	--tls-cert-file <path to your client cert> \
	--tls-key-file <path to your client key> \
	--tls-ca-file <path to your CA> \ 
	--ep localhost -p 5432 -u temporal -pw temporal --pl postgres --db temporal_visibility update-schema -d ./schema/postgresql/v96/visibility/versioned
```

Reviewer: sample commands for mysql:
```zsh
./temporal-sql-tool \
	--tls \
	--tls-enable-host-verification \
	--tls-cert-file <path to your client cert> \
	--tls-key-file <path to your client key> \
	--tls-ca-file <path to your CA> \ 
	--ep localhost -p 3036 -u root -pw root --pl mysql --db temporal update-schema -d ./schema/mysql/v57/temporal/versioned/
```

```zsh
./temporal-sql-tool \
	--tls \
	--tls-enable-host-verification \
	--tls-cert-file <path to your client cert> \
	--tls-key-file <path to your client key> \
	--tls-ca-file <path to your CA> \ 
	--ep localhost -p 3036 -u root -pw root --pl mysql --db temporal_visibility update-schema -d ./schema/mysql/v57/visibility/versioned/
```

---
id: server-dependencies
title: Temporal server dependencies
sidebar_label: Dependencies
---

## Overview

This page details some of the requirements and dependencies needed to build and run an instance of Temporal.

All versions of Temporal require that [Go v1.15+](https://golang.org/dl/) is installed in the environment.

## Dependency matrix

The following table shows which versions of dependencies are supported with a specific version of Temporal.

> Note that the only required dependency is a database, and there are multiple types of databases that are supported. Search is an optional feature and currently only ElasticSearch is supported. Event streaming software as a dependency is only required when ElasticSearch is being used or when Temporal is deployed across multiple data centers, however in future releases of Temporal, third party event streaming software will likely cease to be needed as dependency for both. Temporal emits metrics by default in a format that is supported by Prometheus. Monitoring and observing those metrics is optional. Any software that can pull metrics that supports the same format could be used, but we only ensure it works with Prometheus and Grafana versions.
> 
> Legend of tested and supported dependency versions:
> 
> - 🟩 &nbsp;**Cassandra v3.11**
> - 🟪 &nbsp;**MySQL v5.7**
> - 🟧 &nbsp;**PostgreSQL v9.6**
> - 🟥 &nbsp;**ElasticSearch v6.8**
> - 🟦 &nbsp;**Prometheus >= v2.0**
> - 🟨 &nbsp;**Grafana >= v2.5**
> - 🟫 &nbsp;**Kafka v2.1.1 & Zookeeper v3.4.6**
> 
> 
> | Version | Go packages | Databases | Search | Event streaming | Monitoring | Observation |
> |---------|-------------|-----------|--------|-----------------|------------|-------------|
> | [1.3.2](https://github.com/temporalio/temporal/tree/v1.3.2) | [go.mod](https://github.com/temporalio/temporal/blob/v1.3.2/go.mod) | 🟩 🟪 🟧 | 🟥 | 🟫 | 🟦 | 🟨 |
> | [1.3.1](https://github.com/temporalio/temporal/tree/v1.3.1) | [go.mod](https://github.com/temporalio/temporal/blob/v1.3.1/go.mod) | 🟩 🟪 🟧 | 🟥 | 🟫 | 🟦 | 🟨 |
> | [1.3.0](https://github.com/temporalio/temporal/tree/v1.3.0) | [go.mod](https://github.com/temporalio/temporal/blob/v1.3.0/go.mod) | 🟩 🟪 🟧 | 🟥 | 🟫 | 🟦 | 🟨 |
> | [1.2.2](https://github.com/temporalio/temporal/tree/v1.2.2) | [go.mod](https://github.com/temporalio/temporal/blob/v1.2.2/go.mod) | 🟩 🟪 🟧 | 🟥 | 🟫 | 🟦 | 🟨 |
> | [1.2.1](https://github.com/temporalio/temporal/tree/v1.2.1) | [go.mod](https://github.com/temporalio/temporal/blob/v1.2.1/go.mod) | 🟩 🟪 🟧 | 🟥 | 🟫 | 🟦 | 🟨          |
> | [1.1.1](https://github.com/temporalio/temporal/tree/v1.1.1) | [go.mod](https://github.com/temporalio/temporal/blob/v1.1.1/go.mod) | 🟩 🟪 | 🟥 | 🟫 | 🟦 | 🟨 |
> | [1.1.0](https://github.com/temporalio/temporal/tree/v1.1.0)   | [go.mod](https://github.com/temporalio/temporal/blob/v1.1.0/go.mod) | 🟩 🟪 | 🟥 | 🟫 | 🟦 | 🟨 |
> | [1.0.0](https://github.com/temporalio/temporal/tree/v1.0.0) | [go.mod](https://github.com/temporalio/temporal/blob/v1.0.0/go.mod) | 🟩 🟪 | 🟥 | 🟫 | 🟦 | 🟨 |

Reviewer 2: just a suggestion to reorg info, take or leave as you wish

```suggestion
Temporal offers official support for and is tested against dependencies with these exact versions:

- Databases (Must have one): Note that the only required dependency is a database, and there are multiple types of databases that are supported. 
	- **Cassandra v3.11**
	- **MySQL v5.7**
	- **PostgreSQL v9.6** (supported since Temporal v1.2.1)
- Search (Optional): Search is an optional feature and currently only ElasticSearch is supported. 
	- **ElasticSearch v6.8**
- Monitoring & Observability (Optional): Temporal emits metrics by default in a format that is supported by Prometheus. Monitoring and observing those metrics is optional. Any software that can pull metrics that supports the same format could be used, but we only ensure it works with Prometheus and Grafana versions.
	- **Prometheus >= v2.0**
	- **Grafana >= v2.5**
- Sreaming (Optional, to be deprecated in future): Event streaming software as a dependency is only required when ElasticSearch is being used or when Temporal is deployed across multiple data centers, however in future releases of Temporal, third party event streaming software will likely cease to be needed as dependency for both. 
	- **Kafka v2.1.1 & Zookeeper v3.4.6**
```

All versions of Temporal require that [Go v1.15+](https://golang.org/dl/) is installed in the environment.

## Dependencies

> Temporal offers official support for and is tested against dependencies with these exact versions:

Reviewer 2: ```suggestion
Temporal offers official support for, and is tested against, dependencies with these exact versions:
```

| [1.1.1](https://github.com/temporalio/temporal/tree/v1.1.1) | [go.mod](https://github.com/temporalio/temporal/blob/v1.1.1/go.mod) |
| [1.1.0](https://github.com/temporalio/temporal/tree/v1.1.0) | [go.mod](https://github.com/temporalio/temporal/blob/v1.1.0/go.mod) |
| [1.0.0](https://github.com/temporalio/temporal/tree/v1.0.0) | [go.mod](https://github.com/temporalio/temporal/blob/v1.0.0/go.mod) |

### Databases

The only required dependency is a database, and there are multiple types of databases that are supported.

- **Cassandra v3.11**
- **MySQL v5.7**
- **PostgreSQL v9.6** (supported since Temporal v1.2.1)

### Search

Search is an optional feature and currently only ElasticSearch is supported.

- **ElasticSearch v6.8**

### Monitoring & Observability

Temporal emits metrics by default in a format that is supported by Prometheus. Monitoring and observing those metrics is optional. Any software that can pull metrics that supports the same format could be used, but we only ensure it works with Prometheus and Grafana versions.

- **Prometheus >= v2.0**
- **Grafana >= v2.5**

> ### Sreaming

Reviewer 2: ```suggestion
### Streaming
```

> 	--ep localhost -p 3036 -u root -pw root --pl mysql --db temporal_visibility update-schema -d ./schema/mysql/v57/visibility/versioned/

Reviewer 2: where do i find the docs for `temporal-cassandra-tool`? shall we link people to it

Author: None yet AFAIK -

Reviewer 2: yikes
